### DISCUSSION TITLE
Roadmap for support for partial goals/asynchronous motion

### DISCUSSION THREAD

**Discussion by marip8** [Yaskawa-Global/motoros2#218] [Q&A]
Per the documentation, MotoROS2 supports multiple motion groups, but it does not currently allow them to be moved [independently (i.e., partial goal)](https://github.com/Yaskawa-Global/motoros2?tab=readme-ov-file#no-support-for-partial-goals) or [asynchronously from one another](https://github.com/Yaskawa-Global/motoros2?tab=readme-ov-file#no-support-for-asynchronous-motion).  These features, particularly the partial goals, would be very handy for robot + external axis systems, and it seems like it could be straightforward to implement the proposed workaround (i.e., setting un-specified joints to their current value for the duration of the trajectory) under the hood in the driver itself (at least for partial goals). I believe the ROS1 driver also had the capability of partial goals via the `topic_list` parameter

What is the roadmap for adding these features?

**Comment by ted-miller** [comment]
The next 'major' planned update is to add compatibility for Iron. Beyond that, we're open to requests. 

There's been some activity regarding a [streaming interface](https://github.com/Yaskawa-Global/motoros2/discussions/186). But if there is a use-case for partial goals, I'm open to looking at that first.

How would this type of functionality work in your case? You just want to omit an entire control-group when sending a trajectory? Or are you looking to omit individual axes within a group? Are there any safety concerns since the motion planner isn't controlling everything in the system?

**Reply by marip8** [reply]
> You just want to omit an entire control-group when sending a trajectory? Or are you looking to omit individual axes within a group?

For my use case (robot + 2-axis positioner), I would like to just omit entire groups to be able to control:
- The robot only
- The positioner only
- The robot and positioner together, synchronously

**Comment by gavanderhoorn** [comment]
After writing those FAQs and the roadmap items, I've started to dislike the idea of supporting at least partial goals in MotoROS2.

Main reason is I feel it can lead to unsafe situations 'too easily'.

We have a GP25 on a rail here and in `motoman_driver` it *is* possible to submit partial goals (well, on our system it is anyway). 'Forgetting' to include the rail is scary at best, catastrophic at worst. As you write, the driver basically just copies the 'last known joint state' for the non-included joints.

Asynchronous control of groups could perhaps be a compromise, but would probably require either supporting multiple concurrent Action requests -- targetting different groups -- or multiple action servers.

The former is non-trivial, the latter is likely difficult due to resource constraints.

**Reply by gavanderhoorn** [reply]
If we can come up with a sufficiently convincingly safe way to implement it though (without being overly complex), I would not be opposed to it.

**Reply by marip8** [reply]
> 'Forgetting' to include the rail is scary at best, catastrophic at worst

What do you mean by 'forgetting' specifically?

In my application, I would specifically choose a predefined motion group for ROS motion planning (e.g., robot only, positioner only, or robot + positioner) that corresponds to one of the groups defined on the controller. I would then would submit a trajectory for that specific group with the assumption that the robot controller holds the unspecified joints at their current position.

Is the risk that the user translates a ROS motion plan incorrectly to the joint trajectory action, such that a joint (like your rail) was accidentally not included?

**Reply by gavanderhoorn** [reply]
A motion planner is just one way to generate a `JointTrajectory`.

MoveIt groups, Tesseract groups, anything written by hand can all be misconfigured, but at least we currently have a (simple) way to detect that. If any joint is missing from a goal, it gets refused. Partial goals -- as the term is defined by the `ros_control/JointTrajectoryController` -- would allow trajectories with any subset of joints to be executed.

This makes implementing checks on the action server side rather difficult, as any combination would have to be considered valid.

**Reply by marip8** [reply]
> Partial goals -- as the term is defined by the ros_control/JointTrajectoryController -- would allow trajectories with any subset of joints to be executed.

I see. I agree that allowing any subset of joints to move could be challenging to check and potentially dangerous.

I personally would just like the ability to synchronously command the defined controller motion groups individually (e.g., robot only, positioner only) and collectively with any* combination of controller motion groups (e.g., robot + positioner, left arm + right arm, left arm + torso, left arm + right arm + torso, etc.).

\* I think this is 255 possible motion group combinations for a controller that supports 8 total motion groups:

$$\left( \sum_{k=1}^8 8! / (k!(8-k)!) \right)$$

This doesn't seem like a prohibitive number of possibilities to check, but I would also be okay with copying the current ROS1 strategy of supporting any individual motion group or all motion groups collectively, rejecting all other combinations

Would this be an acceptable compromise?

**Reply by gavanderhoorn** [reply]
I'm not sure I immediately understand the combinatorial in your comment.

In essence, *if* we'd only accept goals for complete motion groups, and we'd have just a single action server, we'd have to check incoming goals and make sure they provide data for the motion group(s) they target.

In other words: if `group_1/joint_1` is in `joint_names`, *all* other joints from `group_1` should be present as well and data must be provided for that group.

If there are no other joints -> only `group_1` would be moved.

If there are any other joints in the goal -> *all* joints from *that* group must also be present in `joint_names` *and* the trajectory must provide data (ie: positions, etc) for that group.

Any goals violating these requirements would get rejected.

This would still only allow a single goal to be active at any given time, but would make it possible to move a subset of motion groups.

---

Technically of course, MotoROS2 could support true partial goals (ie: mix-and-match of joints from all configured motion groups, keeping others at their 'current state').

And at some level, we could say that it's completely up to the user to make sure no unsafe motions are ever submitted.

That might make this easier, as we'd have less to check.

I might be too conservative here, so I'm open to other opinions. @ted-miller @EricMarcil?

**Reply by ted-miller** [reply]
> If there are any other joints in the goal -> all joints from that group must also be present in joint_names and the trajectory must provide data (ie: positions, etc) for that group.
> 
> Any goals violating these requirements would get rejected.

I'm ok with this approach. 

If commanding a group of joints, the user should be explicit about all joints in that group.

> Technically of course, MotoROS2 could support true partial goals (ie: mix-and-match of joints from all configured motion groups, keeping others at their 'current state').

IMO, that's too 'free-for-all'. I know some of the stupid mistakes that I've made when commanding motion. A little bit of restrictions for the sake of safety are a good thing.

**Reply by marip8** [reply]
> I'm not sure I immediately understand the combinatorial in your comment.

I was mostly curious about the total number of combinations that could need to be checked if we took this approach. It's just the sum of an n-choose-k combinatorial where n=8 (8 possible motion groups to command) and k is the number of motion groups being commanded (varying from 1 to 8). All that to say, it's not an astronomically high number of combinations to check

Otherwise, [@gavanderhoorn 's comment](https://github.com/Yaskawa-Global/motoros2/discussions/218#discussioncomment-8705034) summarizes my proposal accurately

**Reply by gavanderhoorn** [reply]
@marip8: would you (or one of your colleagues) perhaps be interested in contributing this feature?

**Reply by marip8** [reply]
Sure; AFAIK we don't have an active seat of the MotoPlus SDK, so we'll have to work through some of the details there. We can do that offline though

**Reply by gavanderhoorn** [reply]
> AFAIK we don't have an active seat of the MotoPlus SDK

in the private beta round, Doug (Smith) did have access to the SDK?

**Reply by marip8** [reply]
He did, but I was told the license was time limited and has since expired. Let me double check though

**Reply by ted-miller** [reply]
@marip8,
We don't offer time-based licenses for the MotoPlus SDK. The license is controlled by a USB dongle. It will look like one of these two, depending on the age:

![image](https://github.com/Yaskawa-Global/motoros2/assets/4623811/e2696cd5-9ffb-4ad1-bc62-c74626818b0e)

![image](https://github.com/Yaskawa-Global/motoros2/assets/4623811/86478b17-696e-4e7b-8a5c-fb9b5d692b87)

**Reply by gavanderhoorn** [reply]
friendly ping @marip8?

**Reply by marip8** [reply]
I found the USB drive. I assume I also need to track down the Windows computer with Visual Studio + MotoPlus SDK to test it, right? Or can I just download the MotoPlus plugin for Visual Studio somewhere?

**Reply by ted-miller** [reply]
I can send you a download link for the latest M+SDK installer. You will need a windows computer.

**Reply by gavanderhoorn** [reply]
Friendly ping @marip8?

**Reply by marip8** [reply]
I haven't been able to look into this much yet, but I did get the download from @ted-miller, but I found out that our M+SDK was purchased/acquired on a client project and I don't have access to it now. I'll follow up offline about getting access to a seat of M+SDK

**Reply by gavanderhoorn** [reply]
Friendly ping again @marip8 :)

**Reply by gavanderhoorn** [reply]
@marip8 ? :)

**Comment by jimmy-mcelwain** [comment]
Copying my comment from #399 (which I probably should have just posted here)

> I looked into this briefly just to see how mpExRcsIncrementMove would work with asynchronous motion. Implementing asynchronous motion will complicate the job setup.
> 
> Asynchronous motion cannot be done with our current INIT_ROS job. I think that we will need a second job, INIT_ROS_ASYNC or something like that. INIT_ROS_ASYNC will need to be a non-group concurrent job that calls PSTART on several jobs, each one being a group job.
> 
> For example, for an R1 + R2 system with asynchronous motion, you would need
> 
> INIT_ROS_ASYNC
> ROS_R1
> ROS_R2
> 
> For R1 + R2 + B1 + B2, you would need
> 
> INIT_ROS_ASYNC
> ROS_R1
> ROS_R2
> ROS_B1
> ROS_B2
> 
> I believe that you could theoretically have some control groups synchronize and others not, e.g.
> 
> INIT_ROS_ASYNC
> ROS_R1_B1
> ROS_R2_B2
> 
> But this would make things more complicated in terms of configuration, inform generation, request parsing, and motion control.
> 
> After submitting trajectories, a call to mpExRcsIncrementMove would be done for each job whose group has data in the incremental motion queue. So in a fully asynchronous R1 + R2 + B1 system where there is data in the queue for both R1 and B1, there would be 2 calls to mpExRcsIncrementMove every 4ms until one of the groups has no data left.
> 
> I am not really sure how to handle some of this. Should we change our FJT action server to have one goal handle for each group? Or do we want 2 action servers, one for synchronous motion and one for asynchronous motion? Do we just want to expose only one FJT action server and have users to pick synchronous or asynchronous through the config file? Or we could have a new service /start_traj_mode_async, or start passing arguments specifying the traj mode into /start_traj_mode. We could also assume that all joints sent in a given FJT message should be synchronized (this probably isn't feasible given the job constraints). It seems like there are several ways to do this, and I don't know what the standard is.
> 
> Everything I've mentioned here is about FJT, but it is mostly true for point queue mode too.
> 
> There are a couple of other motion related ones like https://github.com/Yaskawa-Global/motoros2/issues/51 and especially https://github.com/Yaskawa-Global/motoros2/issues/29 that would be helpful for asynchronous motion, and could maybe be done at the same time as this.

And a reply from @gavanderhoorn 

> An alternative approach which I don't believe you've considered here: internally (ie: internal to MotoROS2) merge incoming overlapping goals for different groups and feed it all as a 'single' trajectory to the task processing / handling the increment queue in the current implementation.
> 
> Groups without an active trajectory would then 'just not move', similar to how it works now.
> 
> It would essentially mean supporting partial goals -- so goals wouldn't have to include trajectories for groups / joints that don't need to move

**Reply by jimmy-mcelwain** [reply]
I have some questions about the merging incoming overlapping goals approach. 

I'll start with this. Right now, we only support 1 goal handle for our action server. In the case of merging overlapping goals, would we increase the number of goal handles so several clients could be waiting for a response? In a situation where R1 is actively executing a motion goal and a R2 motion goal is submitted, would the goal handle for R1 be aborted? Or would it continue to receive feedback for R1 and the client that submitted the new goal would receive feedback for R2? I know we have to consider resource constraints in our design.

**Reply by gavanderhoorn** [reply]
Ideally, I'd like to see `N + 1` action servers (where `N` is the nr of groups on a controller).

One for each control group, and one for all control groups combined. This would mimic the current `JointState` publisher setup, and would be quite clear on *what* a particular FJT server controls, ie: send a goal to the server for a particular group -> control that group.

That could also work for point queuing: essentially we'd have `N + 1` point queues, in the same way as we'd have FJT servers.

> In the case of merging overlapping goals, would we increase the number of goal handles so several clients could be waiting for a response? In a situation where R1 is actively executing a motion goal and a R2 motion goal is submitted, would the goal handle for R1 be aborted? Or would it continue to receive feedback for R1 and the client that submitted the new goal would receive feedback for R2? 

with the configuration I envision, things could keep functioning as they currently do:

- all action servers can only have one active goal at a time
- the `all-groups-combined` server would also check for group overlap and reject if it detects it (ie: `R1` is executing a goal, new goal needs `R1+R2` -> reject)
- per-group servers would also check against `all-groups-combined` (ie: `R1` needed, but `R1+R2` already executing -> reject)

This would allow for asynchronous execution of motion across all groups, as long as the *all-groups* FJT is not already executing a goal.

An alternative approach -- which I believe you had in mind when you wrote your comment -- would be to support multiple concurrent goals as long as they don't need the same groups, while still only having a single action server and supporting partial goals.

I haven't thought about it in detail, but that would seem to make internal bookkeeping rather complex. Instead of essentially 'just' running multiple action servers in parallel (each with their own internal state), we'd have to handle multiple concurrent goals within one instance.

Without prototyping I wouldn't be able to say anything definitive, but I would expect the latter to introduce more complexity than the former.

---

Edit: thinking more about it: we could also opt to change the behaviour of the FJT action server to align more with the implementation in `ros(2)_control`:

- submitting a goal while one is active causes the controller to *replace* the active trajectory with the new one (either after a full stop or using some form of blending (#51))
- submitting a goal with an *empty* trajectory cancels the current goal
- cancelling an active goal stops processing (we already do this)

This wouldn't change what I wrote above fundamentally, except we'd not be rejecting goals when one is active already.

I might suggest doing a (limited form of a) risk analysis to see what would result in the safest default behaviour.

**Reply by jimmy-mcelwain** [reply]
This all sounds reasonable to me. Having `N + 1` action servers sounds good to me if we take that approach. 

Just getting some ideas down about potential roadblocks:
1. Our FSU correction code is going to be a complicating factor. If we take the "merging" approach at its most basic, then if one control group slows down because it enters FSU bounds, then all control groups would slow down to match. I think that the FSU could also just fully stop one group, which would mean all groups would get stopped. We could probably add some logic to avoid this (I haven't thought about it too much), but the FSU correction code is already pretty complicated. This is one area that the multiple jobs/truly asynchronous approach might make more sense. 
2. This may be a good thing for safety reasons, but a failed call to `mpExRcsIncrementMove` would mean that all control groups fail/stop. So if `R1` and `R2` are executing asynchronously and `R2` fails, then `R1` would stop, so they are not "truly" asynchronous. 

Overall I think that it probably makes more sense to have the merging approach that you mentioned, but I'm just trying to think of some of the differences/drawbacks.

**Reply by gavanderhoorn** [reply]
To clarify: `N + 1` action servers doesn't necessarily mean `N + 1` increment queue consumers. It might actually be beneficial to have a single thread responsible for queue processing and calling `mpExRcsIncrementMove(..)` after having 'gathered' all the increments for all active groups and setting up and populating the `MP_EXPOS_DATA` structure.

In a way, there could be `N + 1` *front-ends*, with only a single *back-end*.

If we refactor things this way, adding real-time streaming and queuing would be just-another-front-end.

> [..] but I'm just trying to think of some of the differences/drawbacks.

that's exactly the purpose of this thread, so keep doing that.

**Reply by jimmy-mcelwain** [reply]
Yeah. `mpExRcsIncrementMove` can actually only be called from one thread (the thread with `MP_INTERPOLATION_CLK` priority). You can call it multiple times in that thread in a single time increment as long as the control groups for the different calls don't overlap. So no matter what, the data needs to be consumed in a single thread. 

Looking at the code a little bit more, I actually don't think that the FSU speed limiter would cause any problems.
